### PR TITLE
refine structure

### PR DESCRIPTION
**Pull request by sowu880**

### CODE REVIEW

File: DICOM/src/Microsoft.Health.Dicom.Anonymizer.Core/AnonymizerRuleHandlers.cs [hunk 1]
```diff
@@ -4,29 +4,32 @@
...
 using Microsoft.Health.Dicom.Anonymizer.Core.Exceptions;
 using Microsoft.Health.Dicom.Anonymizer.Core.Model;
+using Microsoft.Health.Dicom.Anonymizer.Core.Processors;
 using Microsoft.Health.Dicom.Anonymizer.Core.Processors.Settings;
 using Newtonsoft.Json.Linq;
 
 namespace Microsoft.Health.Dicom.Anonymizer.Core.AnonymizerConfigurations
 {
-    public class AnonymizerDicomTagRule
+    public class AnonymizerRuleHandlers
```
Comment (qiwjin) [inline DICOM/src/Microsoft.Health.Dicom.Anonymizer.Core/AnonymizerRuleHandlers.cs:19]: wired class name

File: DICOM/src/Microsoft.Health.Dicom.Anonymizer.Core/Processors/CryptoHashProcessor.cs [hunk 2]
```diff
@@ -19,14 +19,14 @@ namespace Microsoft.Health.Dicom.Anonymizer.Core.Processors
...
 
-        public CryptoHashProcessor(DicomCryptoHashSetting defaultSetting = null)
+        public CryptoHashProcessor(IDicomAnonymizationSetting ruleSetting = null)
         {
-            _defaultSetting = defaultSetting ?? new DicomCryptoHashSetting();
+            _ruleSetting = (DicomCryptoHashSetting)(ruleSetting ?? new DicomCryptoHashSetting());
         }
 
-        public void Process(DicomDataset dicomDataset, DicomItem item, DicomBasicInformation basicInfo = null, IDicomAnonymizationSetting settings = null)
+        public void Process(DicomDataset dicomDataset, DicomItem item, ProcessContext context = null)
```
Comment (qiwjin) [inline DICOM/src/Microsoft.Health.Dicom.Anonymizer.Core/Processors/CryptoHashProcessor.cs:29]: need both DicomDataset and item?

File: DICOM/src/Microsoft.Health.Dicom.Anonymizer.Core/Rules/AnonymizerRuleFactory.cs [hunk 3]
```diff
@@ -4,68 +4,38 @@
...
-        public string Method { get; set; }
-
-        public bool IsVRRule { get; set; } = false;
-
-        public bool IsMasked { get; set; } = false;
-
-        public IDicomAnonymizationSetting RuleSetting { get; set; }
-
-        public static AnonymizerDicomTagRule CreateAnonymizationDicomRule(JObject rule, AnonymizerConfiguration configuration)
+        public AnonymizerRule CreateAnonymizationDicomRule(JObject rule)
```
Comment (qiwjin) [inline DICOM/src/Microsoft.Health.Dicom.Anonymizer.Core/Rules/AnonymizerRuleFactory.cs:36]: change rule to object

File: DICOM/src/Microsoft.Health.Dicom.Anonymizer.Core/AnonymizerEngine.cs [hunk 4]
```diff
@@ -3,44 +3,61 @@
...
 
-        public void AnonymizeDateset(DicomDataset dataset)
+        public void AnonymizeDataset(DicomDataset dataset)
         {
+            var context = InitContext(dataset);
             ValidateInput(dataset);
             dataset.AutoValidate = _anonymizerSettings.AutoValidate;
-            _ruleHandler.Handle(dataset);
+
+            var rules = _configurationManager.GetConfiguration().DicomRules?.Select(entry => _ruleFactory.CreateAnonymizationDicomRule(entry)).ToArray();
```
[the remark this answers is not in the document]
Comment (sowu880) [reply DICOM/src/Microsoft.Health.Dicom.Anonymizer.Core/AnonymizerEngine.cs]: Fixed it.

File: DICOM/src/Microsoft.Health.DeID.SharedLib/CryptoHashFunction.cs [hunk 5]
```diff
@@ -13,62 +13,69 @@ namespace Microsoft.Health.Dicom.DeID.SharedLib
...
             {
                 return input;
             }
 
             HMAC hmac = new HMACSHA256();
-            if (hashKey != null)
+            if (_hashKey != null)
             {
-                hmac = new HMACSHA256(hashKey);
+                hmac = new HMACSHA256(_hashKey);
```
Comment (qiwjin) [inline DICOM/src/Microsoft.Health.DeID.SharedLib/CryptoHashFunction.cs:33]: move to constructor

File: DICOM/src/Microsoft.Health.Dicom.Anonymizer.Core/Rules/AnonymizerRule.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+            var locatedItems = LocateDicomTag(dataset, context);
+
+            foreach (var item in locatedItems)
+            {
+                Processor.Process(dataset, item, context);
+                context.VisitedNodes.Add(item.Tag.ToString());
+            }
+        }
+
+        public virtual List<DicomItem> LocateDicomTag(DicomDataset dataset, ProcessContext context)
```
Comment (qiwjin) [inline DICOM/src/Microsoft.Health.Dicom.Anonymizer.Core/Rules/AnonymizerRule.cs:37]: change to abstract

**Review by qiwjin — APPROVED** [review]

**Pull request merged**